Merge X.L.LayoutCombinators.(|||) into X.L.(|||)

slotThe opened the pull request:
### Description

The functionality of the forme are quite handy to have in core and we
can do so with minimal code changes.

The drawbacks of this approach are

  1. We can't merge `JumpToLayout` into `ChangeLayout` because people may
     have imported `JumpToLayout(..)`, which we can't simulate with type
     aliases

  2. Because the internal structure of `X.L.LayoutCombintors.(|||)` is a
     bit different, this creates a regression for people who used
     `NextLayoutNoWrap` or `Wrap` in their configs.  We could work around
     this by creating fake instances of these fields in the new
     `JumpToLayout` constructor and simply not doing anything with them,
     but since this seems like quite an advanced and specific use-case,
     failing fast and hard (as opposed to adding deprecation messages
     and then "silently" not handling these messages) seems preferrable
     here.

### Checklist

  - [X] I've read [CONTRIBUTING.md](https://github.com/xmonad/xmonad/blob/master/CONTRIBUTING.md)

  - [X] I've confirmed these changes don't belong in xmonad-contrib instead

  - [ ] I tested my changes with [xmonad-testing](https://github.com/xmonad/xmonad-testing)

  - [X] I updated the `CHANGES.md` file

slotThe committed:
Merge X.L.LayoutCombinators.(|||) into X.L.(|||)

The functionality of the former are quite handy to have in core and we
can do so with minimal code changes.

The drawbacks of this approach are

  1. We can't merge JumpToLayout into ChangeLayout because people may
     have imported JumpToLayout(..), which we can't simulate with type
     aliases, patterns, and the like.

  2. Because the internal structure of X.L.LayoutCombintors.(|||) is a
     bit different, this creates a regression for people who used
     NextLayoutNoWrap or Wrap in their configs.  We could work around
     this by creating fake instances of these fields in the new
     JumpToLayout constructor and simply not doing anything with them,
     but since this seems like quite an advanced and specific use-case,
     failing fast and hard (as opposed to adding deprecation messages
     and then "silently" not handling these messages) seems preferrable
     here.

Related: https://github.com/xmonad/xmonad-contrib/pull/493
Related: https://github.com/xmonad/xmonad-contrib/issues/116

slotThe committed:
Update CHANGES.md

geekosaur commented:
Speling, for what it's worth :)

slotThe commented:
Whoops :)

geekosaur commented on CHANGES.md:
```diff
@@ -35,6 +35,10 @@
   * Fixed dunst notifications being obscured when moving floats.
     https://github.com/xmonad/xmonad/issues/208
 
+  * Migrated `X.L.LayoutCombinator.(|||)` into `XMonad.Layout`,
```
`LayoutCombinators`

geekosaur reviewed (approved):
LGTM.

liskin reviewed (approved):
This is way less messy than I expected. 👍
Sorry for the delay.

Does this need https://github.com/xmonad/xmonad-contrib/pull/493/ to be merged to avoid breakage or is that PR optional? Asking just to understand the dependencies and migration path.

liskin commented:
Anyway, please mention https://github.com/xmonad/xmonad-contrib/pull/493 in the commit message :-)

slotThe commented:
> Does this need [xmonad/xmonad-contrib#493](https://github.com/xmonad/xmonad-contrib/pull/493) to be merged to avoid breakage or is that PR optional? Asking just to understand the dependencies and migration path.

Yes, contrib doesn't compile with this change and without xmonad/xmonad-contrib#493; X.L.LayoutCombinators also defines a `JumpToLayout` constructor and then GHC can't figure out which one we mean.  

I'll rebase and mention that pr in the commit message

liskin commented:
> Yes, contrib doesn't compile with this change and without [xmonad/xmonad-contrib#493](https://github.com/xmonad/xmonad-contrib/pull/493); X.L.LayoutCombinators also defines a `JumpToLayout` constructor and then GHC can't figure out which one we mean.

Oh, looks like top-level definitions are in the same lexical scope as imported stuff, I didn't know that.

In that case I'd like to release X11 first, and only then do a "big bang" where we bump versions and dep bounds and merge all the build breaking stuff.

liskin commented on src/XMonad/Layout.hs:
```diff
@@ -155,6 +190,18 @@ instance Message NextNoWrap
 handle :: (LayoutClass l a, Message m) => l a -> m -> X (Maybe (l a))
 handle l m = handleMessage l (SomeMessage m)
 
+-- | Swap the two sub-layouts.
```
This is misleading. I'd expect a `Choose l r a → Choose r l a`, but instead it's a `switch` that also happens to notify the inner layouts that they're going to be hidden.

liskin replied:
Oh and I just noticed that there's this `choose` function that could possibly replace both `swap` and `switch`, as it seems to be more general. IOW, `JumpToLayout` handling may be possible to implement concisely using `choose` alone.

liskin replied:
(If you want to test this, I expect that any decorated layout would fail to destroy its decoration windows if Hide messages aren't delivered to it on time.)

liskin reviewed (changes requested):
Let's try to simplify the impl and merge asap to have at least a few weeks of testing.

slotThe commented:
> Let's try to simplify the impl and merge asap to have at least a few weeks of testing.

Sure, I'll give that a try tomorrow

slotThe committed:
X.Layout: Simplify JumpToLayout implementation

The `choose` combinator is very general; much more so than the
combination of `switch`, `swap`, and `passOnM`.

We can thus replace most of the implementation with calls to `choose`
and properly put everything related to `JumpToLayout` under one guard.
Plus, we don't need to define any extra functions that would have to be
tested in some way.

Co-authored-by: Tomas Janousek <tomi@nomi.cz>

slotThe commented:
Turns out that you were right!  I have pushed 246600f70d6365c612ea9bfd31efd50c50a6c861, which works for me (so fingers crossed it does for others as well :))

liskin commented on src/XMonad/Layout.hs:
```diff
@@ -247,22 +230,19 @@ instance (LayoutClass l a, LayoutClass r a) => LayoutClass (Choose l r) a where
+                  then choose c CR Nothing Nothing
+                  else handleMessage l m >>= \case
+                      Just ml -> choose c CL (Just ml) Nothing
+                      Nothing -> choose (Choose CR l r) CR Nothing =<< handleMessage r m
```
Why do you manually construct `(Choose CR l r)` instead of using `c` here? I think that will suppress the `Hide` events.

slotThe replied:
`c` is of the form `Choose CL l r`, if I wanted to use it I'd need another `switch`-like function.  Putting `c` there would hide the current layout in every case, even if we're on, say, `Full` and executed `sendMessage (JumpToLayout Full)`

liskin replied:
If we're on `Full` and we `JumpToLayout Full`, it's handled by the `desc == description r` case above, isn't it?

`choose (Choose CR l r) CR` and `choose (Choose CL l r) CL` will never `hide` anything, and that doesn't sound right. The `l` layout does need hiding when we're switching away from it, I think.

slotThe replied:
This all makes perfect sense, yet it still works (even with decorated layouts)...  I'm very confused.

We can, however, at least cover the "we're on the same layout" part by checking for descriptions in either case.  This, like magic, also enables us to get rid of the non-hiding calls to `choose` ( 0084a1af5c08fa3015af5aa2fd45d0c165d86a67 )

liskin replied:
Can you explain what the problem was with the "we're on the same layout" part? That may have been the bit I missed…

slotThe replied:
> Can you explain what the problem was with the "we're on the same layout" part? That may have been the bit I missed…

You said that 

  > If we're on `Full` and we `JumpToLayout Full`, it's handled by the `desc == description r` case above, isn't it?

but it might well be that our layouts looks like `Choose CL Full (*other-layouts*)` and so we'd have to check `desc == description l` in the `CL` case as well (which we and also the old `X.L.LayoutCombinators.(|||)` didn't do).  As it stood, this would have then gone through to the other cases, which then managed to switch the layout (I imagine `handleMessage l m` returned `Nothing` and then we flipped `CL` to `CR` in `choose`)

liskin commented on src/XMonad/Layout.hs:
```diff
@@ -230,16 +230,13 @@ instance (LayoutClass l a, LayoutClass r a) => LayoutClass (Choose l r) a where
-                   Nothing -> (\ml -> choose c CL ml Nothing) =<< handleMessage l m
+    handleMessage c@(Choose d l r) m | Just (JumpToLayout desc) <- fromMessage m = do
+        ml <- handleMessage l m
+        mr <- handleMessage r m
```
I should probably not even mention it to you, as we all know what's gonna happen, but still:

This simple implementation always walks the whole tree of Chooses and then possibly passes the message all the way down layout modifiers as well. The previous one stopped descending when it found a node with the desired description, but it still passed the message to some subtrees (all the way to layout modifiers and base layout). Asymptotically it's the same, in practice it might be a couple nanosecs or microsecs difference, I wouldn't worry about it at all. But it can be dealt with by keeping the first two lines of the multi-way if in here.

slotThe replied:
I will go against expectations and prefer clarity to this spurious efficiency gain in this situation then :P
